src/box/tuple.h

```diff
@@ -440,18 +470,39 @@ tuple_is_unreferenced(struct tuple *tuple)
 	return tuple->local_refs == 0;
 }
 
+/** Check that the tuple is in compact mode. */
+static inline bool
+tuple_is_compact(struct tuple *tuple)
+{
+	return tuple->data_offset_bsize_raw & 0x8000;
+}
+
 /** Offset to the MessagePack from the beginning of the tuple. */
 static inline uint16_t
 tuple_data_offset(struct tuple *tuple)
 {
-	return tuple->data_offset;
+	uint16_t res = tuple->data_offset_bsize_raw;
+	res = (res & 0x7fff) >> ((res >> 15) * 8);
```

Comment:
After looking at this line for a few minutes I managed to figure out what's going on here, but I didn't quite like the experience, to be honest. Please do something about this code so that a layman can understand what's going on here without screwing their brains up or add a good comment.

Reply:
OK, added a comment.